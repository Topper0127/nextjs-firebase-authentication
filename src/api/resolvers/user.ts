import { combineResolvers } from 'graphql-resolvers';

import { ResolverContext } from '@typeDefs/resolver';
import { isAuthenticated } from '@api/authorization/isAuthenticated';

export default {
  Query: {
    me: combineResolvers(
      isAuthenticated,
      (parent, args, { me }: ResolverContext) => {
        return {
          email: me.email,
          uid: me.uid,
        };
      }
    ),
  },
};
