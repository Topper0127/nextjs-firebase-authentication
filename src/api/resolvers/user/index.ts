import {
  ObjectType,
  Field,
  Ctx,
  Resolver,
  Query,
  UseMiddleware,
} from 'type-graphql';

import { ResolverContext } from '@typeDefs/resolver';
import { isAuthenticated } from '@api/middleware/resolver/isAuthenticated';

@ObjectType()
class User {
  @Field()
  uid: string;

  @Field()
  email: string;

  @Field()
  username: string;

  @Field(type => [String])
  roles: string[];
}

@Resolver()
export default class UserResolver {
  @Query(() => User)
  @UseMiddleware(isAuthenticated)
  async me(@Ctx() ctx: ResolverContext): Promise<User> {
    const rolesObject = ctx.me!.customClaims || {};

    const roles = Object.keys(rolesObject).filter(
      key => rolesObject[key]
    );

    return {
      uid: ctx.me!.uid,
      email: ctx.me!.email || '',
      username: ctx.me!.displayName || '',
      roles,
    };
  }
}
