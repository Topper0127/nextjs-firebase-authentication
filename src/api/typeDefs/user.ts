import {
  makeExecutableSchema,
  addMockFunctionsToSchema,
  mergeSchemas,
} from 'graphql-tools';
import { gql } from 'apollo-server-micro';

export default makeExecutableSchema({
  typeDefs: `
  type Query {
    me: User
  }

  type User {
    email: String!
    uid: String!
  }
`,
});
