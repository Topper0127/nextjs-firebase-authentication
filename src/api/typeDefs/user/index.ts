import { gql } from 'apollo-server-micro';

export default gql`
  extend type Query {
    me: User
  }

  type User {
    email: String!
    uid: String!
  }
`;
