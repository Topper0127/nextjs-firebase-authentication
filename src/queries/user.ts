import gql from 'graphql-tag';

export const GET_ME = gql`
  query GetMe {
    me {
      uid
      email
      username
      roles
    }
  }
`;
