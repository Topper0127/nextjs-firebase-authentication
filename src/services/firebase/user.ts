import { db } from './firebase';

export const doCreateUser = (
  id: string,
  username: string,
  email: string
) =>
  db.ref(`users/${id}`).set({
    username,
    email,
  });
