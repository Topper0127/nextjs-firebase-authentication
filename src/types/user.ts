import * as firebaseAdminVanilla from 'firebase-admin';

export type User = firebaseAdminVanilla.auth.UserRecord;
