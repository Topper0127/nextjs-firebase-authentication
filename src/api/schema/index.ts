import { mergeSchemas } from 'graphql-tools';
import { applyMiddleware } from 'graphql-middleware';

import { Resolvers } from '@generated/gen-types';

import authorization from '@api/authorization';
import typeDefs from '@api/typeDefs';
import resolvers from '@api/resolvers';

const schema = mergeSchemas({
  schemas: typeDefs,
  resolvers: resolvers as Resolvers,
});

export default applyMiddleware(schema, authorization);
