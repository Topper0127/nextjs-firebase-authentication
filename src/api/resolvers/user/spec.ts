import { createTestClient } from 'apollo-server-testing';
import { ApolloServer } from 'apollo-server-micro';

import schema from '@api/schema';

import { GET_ME } from '@queries/user';

describe('me', () => {
  it('returns the signed in user', async () => {
    const server = new ApolloServer({
      schema,
      context: () => ({
        me: { uid: '1', email: 'example@example.com' },
      }),
    });

    const { mutate } = createTestClient(server);

    const { data, errors } = await mutate({
      mutation: GET_ME,
    });

    expect(data).toMatchSnapshot();
    expect(errors).toEqual(undefined);
  });

  it('returns nothing if no user is signed in', async () => {
    const server = new ApolloServer({
      schema,
      context: () => ({
        me: null,
      }),
    });

    const { mutate } = createTestClient(server);

    const { data, errors } = await mutate({
      mutation: GET_ME,
    });

    expect(data).toMatchSnapshot();
    expect(errors).toMatchSnapshot();
  });
});
