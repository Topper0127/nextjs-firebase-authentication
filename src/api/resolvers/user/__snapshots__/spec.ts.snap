// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`me returns nothing if no user is signed in 1`] = `
Object {
  "me": null,
}
`;

exports[`me returns nothing if no user is signed in 2`] = `
Array [
  [GraphQLError: Not authenticated as user.],
]
`;

exports[`me returns the signed in user 1`] = `
Object {
  "me": Object {
    "email": "example@example.com",
    "uid": "1",
  },
}
`;
